Silence the "DSP panic!" prints during firmware boot on HDA

ujfalusi opened the pull request:
Hi,

As it has been requested we should not print the panic information during firmware boot (on HDA platforms) if the `SOF_DBG_PRINT_ALL_DUMPS` is not set.

plbossart commented on sound/soc/sof/intel/hda-ipc.c:
```diff
@@ -173,8 +173,21 @@ irqreturn_t hda_dsp_ipc_irq_thread(int irq, void *context)
+			 * this is a PANIC message !!
+			 * If it is arriving during firmware boot and it is not
+			 * the last boot attempt then mark the no_recovery as
+			 * false to indicate that we be able to boot up
```
missing word: we be able to boot up

did you mean:

the DSP might be able to boot in the next iteration(s).

ujfalusi replied:
"the DSP might be able to boot in the next iteration(s).", yes I meant this.

ujfalusi replied:
Done

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -82,9 +82,22 @@ irqreturn_t cnl_ipc_irq_thread(int irq, void *context)
-			snd_sof_dsp_panic(sdev, HDA_DSP_PANIC_OFFSET(msg_ext));
+		if ((hipctdr & SOF_IPC_PANIC_MAGIC_MASK) == SOF_IPC_PANIC_MAGIC) {
+			struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+			bool no_recovery = true;
```
I would use 'non_recoverable' - a qualifier to the panic

'no_recovery' seems to hint that we would disable a TBD recovery mechanism, that's not at all what this is about.

same comments for the hda-ipc code below.

ujfalusi replied:
Or we could use 'last_boot_attempt'?

This all comes from my (bad?) choice of the `bool non_recoverable` new parameter to snd_sof_dsp_panic()...
The aim was to have a **true** passed in case this is a 'real' and 'definite' panic and a **false** would indicate that the platform code might be able to deal with it, but this is still a panic from the firmware and if user wants to see _all_ dumps then it should be shown.

plbossart replied:
'last_boot_attempt' sounds good, it is exactly what we want to say :-)

ujfalusi replied:
Hrm, `last_boot_attempt` is true for the period when we are booting up but after that if the panic comes when we have already booted up, it is a bit strange.
Probably the `non_recoverable` is a bit better match for the cases we handle here.

ranj063 commented on sound/soc/sof/ops.c:
```diff
@@ -144,18 +144,22 @@ EXPORT_SYMBOL(snd_sof_dsp_update_bits_forced);
+	 */
+	if (sdev->dsp_oops_offset != offset)
+		dev_warn(sdev->dev, "%s: dsp_oops_offset %zu offset %u\n", __func__,
+			 sdev->dsp_oops_offset, offset);
```
can we add why it is a warning? i.e dsp_oopsoffset %zu differs from panic offset %u?

ujfalusi replied:
Yes, certainly a good idea.

ranj063 commented on sound/soc/sof/core.c:
```diff
@@ -27,6 +27,22 @@ MODULE_PARM_DESC(sof_debug, "SOF core debug options (0x0 all off)");
+ */
+bool sof_debug_check_flag(int mask)
+{
+	if ((sof_core_debug & mask) == mask)
```
this makes sense if we want to check multiple flags at once but in most cases we'd be checking if a single mask is set no? so shouldnt if (sof_core_debug & mask) suffice?

ujfalusi replied:
Currently it is true, but while here I thought why not do it in a way that might be usable for the future, if ever someone needs to check for two set bits at the same time.

I'll leave it like this for now, but if more votes for the simple single bit check then I can change to that.

ranj063 replied:
@ujfalusi but the problem is if we do have more than 1 flag set in sof_core_debug, you need to make sure to check all the flags for this to be correct no?

ujfalusi replied:
@ranj063, I'm not following you. With this you can check if the bits set in the mask are also set in the sof_core_debug and all of the bits are set.
sof_debug_check_flag(BIT(0) | BIT(2)); will only return true if both bit 0 _and_ bit 2 is set in sof_core_debug
sof_debug_check_flag(BIT(0)) will return true if bit 0 is set in sof_core_debug
sof_debug_check_flag(BIT(2)) will return true if bit 2 is set in sof_core_debug

ranj063 replied:
sorry @ujfalusi  I misunderstood the mask. ignore my comment. i'm good with this.

ujfalusi commented:
Changes since v1:
- no_recovery changed to non_recoverable in hda-ipc.c and cnl.c
 - comment section is updated as well
- Add the reason why we print a warn for the offset differences.

ujfalusi commented:
Changes since v2:
- New patch to add a 'message' parameter to snd_sof_dsp_dbg_dump() to handle the DSP dump reason print.

ujfalusi commented:
Changes since v3:
- checkpatch fix (caused by copy-paste and neglect)

ranj063 commented on sound/soc/sof/debug.c:
```diff
@@ -964,12 +964,17 @@ void snd_sof_dsp_dbg_dump(struct snd_sof_dev *sdev, u32 flags)
 		if (!print_all)
 			sdev->dbg_dump_printed = true;
+	} else if (msg) {
+		dev_err(sdev->dev, "%s\n", msg);
```
is this else part needed?

ujfalusi replied:
Even if we are not going to show the full dump, we want to show the message. As it is done currently.

ranj063 replied:
ahh I see. you've moved the error messages too, nice!

ujfalusi replied:
The idea is that we will have it within the marks and in the same location for all dumps, thus making it easier to parse.

ranj063 commented on sound/soc/sof/debug.c:
```diff
@@ -964,12 +964,17 @@ void snd_sof_dsp_dbg_dump(struct snd_sof_dev *sdev, u32 flags)
+	} else if (msg) {
+		dev_err(sdev->dev, "%s\n", msg);
 	}
+
```
spurious change?

ujfalusi replied:
oops.

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -189,9 +190,12 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
-	snd_sof_dsp_dbg_dump(sdev, flags);
+	dump_msg = kasprintf(GFP_KERNEL, "Boot iteration failed: %d/%d",
+			     hda->boot_iteration, HDA_FW_BOOT_ATTEMPTS);
+	snd_sof_dsp_dbg_dump(sdev, dump_msg, flags);
```
but do we want to print the error if we're at < HDA_FW_BOOT_ATTEMPTS?

ujfalusi replied:
We are not going to if the `SOF_DBG_PRINT_ALL_DUMPS` is not set.
The `SOF_DBG_DUMP_OPTIONAL` flag will skip all prints.

ranj063 replied:
ahh ok. You're probably the only one you understands all these flags really well now :)

ujfalusi commented:
Changes since v4:
- Removed spurious line in debug.c
